nav2_costmap_2d/plugins/static_layer.cpp

```diff
@@ -56,185 +58,198 @@ using nav2_costmap_2d::FREE_SPACE;
 
-  int temp_lethal_threshold, temp_unknown_cost_value;
-  node_->get_parameter_or<int>("lethal_cost_threshold", temp_lethal_threshold, 100);
-  node_->get_parameter_or<int>("unknown_cost_value", temp_unknown_cost_value, -1);
-  node_->get_parameter_or<bool>("trinary_costmap", trinary_costmap_, true);
-
-  lethal_threshold_ = std::max(std::min(temp_lethal_threshold, 100), 0);
-  unknown_cost_value_ = temp_unknown_cost_value;
-
-  // we'll subscribe to the latched topic that the map server uses
-  RCLCPP_INFO(node_->get_logger(), "Requesting the map...");
-  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_default;
-  custom_qos_profile.depth = 1;
-  custom_qos_profile.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
-  map_sub_ = node_->create_subscription<nav_msgs::msg::OccupancyGrid>(map_topic,
-      std::bind(&StaticLayer::incomingMap, this, std::placeholders::_1), custom_qos_profile);
-  map_received_ = false;
-  has_updated_data_ = false;
-
-  rclcpp::Rate r(10);
-  rclcpp::executors::SingleThreadedExecutor exec;
-  while (!map_received_ && rclcpp::ok()) {
-    exec.spin_node_once(node_->get_node_base_interface(), std::chrono::milliseconds(100));
-    r.sleep();
+  if (!first_map_only_) {
+    RCLCPP_INFO(node_->get_logger(), "Subscribing to the map topic");
+    map_sub_ = node_->create_subscription<nav_msgs::msg::OccupancyGrid>(map_topic_,
+      std::bind(&StaticLayer::incomingMap, this, std::placeholders::_1));
   }
 
-  RCLCPP_INFO(node_->get_logger(),
-    "Received a %d X %d map at %f m/pix", getSizeInCellsX(),
-    getSizeInCellsY(), getResolution());
-
   if (subscribe_to_updates_) {
     RCLCPP_INFO(node_->get_logger(), "Subscribing to updates");
     map_update_sub_ = node_->create_subscription<map_msgs::msg::OccupancyGridUpdate>(
-      map_topic + "_updates",
-      std::bind(&StaticLayer::incomingUpdate, this, std::placeholders::_1), custom_qos_profile);
-
-  } else {
-    has_updated_data_ = true;
+      map_topic_ + "_updates",
+      std::bind(&StaticLayer::incomingUpdate, this, std::placeholders::_1));
   }
 
-  dynamic_param_client_ = std::make_unique<nav2_dynamic_params::DynamicParamsClient>(node_);
-  dynamic_param_client_->add_parameters({name_ + "." + "enabled"});
-  dynamic_param_client_->set_callback(std::bind(&StaticLayer::reconfigureCB, this));
-  // TODO(bpwilcox): Add new parameters to parameter validation class from plugins
+  current_ = true;
 }
 
-void StaticLayer::reconfigureCB()
+void
+StaticLayer::activate()
 {
-  RCLCPP_DEBUG(node_->get_logger(), "StaticLayer:: Event Callback");
+}
 
-  bool enabled = true;
-  dynamic_param_client_->get_event_param_or(name_ + "." + "enabled", enabled, true);
+void
+StaticLayer::deactivate()
+{
+}
 
-  if (enabled != enabled_) {
-    enabled_ = enabled;
-    has_updated_data_ = true;
-    x_ = y_ = 0;
-    width_ = size_x_;
-    height_ = size_y_;
-  }
+void
+StaticLayer::reset()
+{
+  map_sub_.reset();
+  map_update_sub_.reset();
+
+  onInitialize();
```

Comment:
From what I can tell, if `reset()` is called (for instance during a recovery), the `onInitialize()` will try call `getParameters()`, which should cause an error since the parameters have already been declared on the node. We may need to undeclare those parameters before calling `onInitialize()` or otherwise cleanup the plugins

Reply:
@bpwilcox That's true. How about I enter a separate issue for this? We need to scrub the plugins for  reinitialization and the proper interface to the lifecycle node control flow. Do you want to take this on? We can do it once we're finished with the integration into the lifecycle branch.

Reply:
I could take it on, though we should only integrate this PR if we are not using/disabling recovery behaviors (particularly clearing costmap) for the short term, otherwise, we'll run into runtime errors in our lifecycle branch every time there's a recovery.